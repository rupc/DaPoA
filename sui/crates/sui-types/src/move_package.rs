// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    base_types::{ObjectID, SequenceNumber},
    error::{ExecutionError, ExecutionErrorKind, SuiError, SuiResult},
};
use move_binary_format::access::ModuleAccess;
use move_binary_format::binary_views::BinaryIndexedView;
use move_binary_format::file_format::CompiledModule;
use move_binary_format::normalized;
use move_core_types::{account_address::AccountAddress, identifier::Identifier};
use move_disassembler::disassembler::Disassembler;
use move_ir_types::location::Spanned;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use serde_with::serde_as;
use serde_with::Bytes;
use std::collections::BTreeMap;

// TODO: robust MovePackage tests
// #[cfg(test)]
// #[path = "unit_tests/move_package.rs"]
// mod base_types_tests;

#[derive(Clone, Debug)]
/// Additional information about a function
pub struct FnInfo {
    /// If true, it's a function involved in testing (`[test]`, `[test_only]`, `[expected_failure]`)
    pub is_test: bool,
}

#[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
/// Uniquely identifies a function in a module
pub struct FnInfoKey {
    pub fn_name: String,
    pub mod_addr: AccountAddress,
}

/// A map from function info keys to function info
pub type FnInfoMap = BTreeMap<FnInfoKey, FnInfo>;

// serde_bytes::ByteBuf is an analog of Vec<u8> with built-in fast serialization.
#[serde_as]
#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
pub struct MovePackage {
    id: ObjectID,
    /// Most move packages are uniquely identified by their ID (i.e. there is only one version per
    /// ID), but the version is still stored because one package may be an upgrade of another (at a
    /// different ID), in which case its version will be one greater than the version of the
    /// upgraded package.
    ///
    /// Framework packages are an exception to this rule -- all versions of the framework packages
    /// exist at the same ID, at increasing versions.
    ///
    /// In all cases, packages are referred to by move calls using just their ID, and they are
    /// always loaded at their latest version.
    version: SequenceNumber,
    // TODO use session cache
    #[serde_as(as = "BTreeMap<_, Bytes>")]
    module_map: BTreeMap<String, Vec<u8>>,
}

impl MovePackage {
    pub fn new(
        id: ObjectID,
        version: SequenceNumber,
        module_map: &BTreeMap<String, Vec<u8>>,
        max_move_package_size: u64,
    ) -> Result<Self, ExecutionError> {
        let pkg = Self {
            id,
            version,
            module_map: module_map.clone(),
        };
        let object_size = pkg.size() as u64;
        if object_size > max_move_package_size {
            return Err(ExecutionErrorKind::MovePackageTooBig {
                object_size,
                max_object_size: max_move_package_size,
            }
            .into());
        }
        Ok(pkg)
    }

    pub fn from_module_iter<T: IntoIterator<Item = CompiledModule>>(
        version: SequenceNumber,
        iter: T,
        max_move_package_size: u64,
    ) -> Result<Self, ExecutionError> {
        let mut iter = iter.into_iter().peekable();
        let id = ObjectID::from(
            *iter
                .peek()
                .expect("Tried to build a Move package from an empty iterator of Compiled modules")
                .self_id()
                .address(),
        );

        Self::new(
            id,
            version,
            &iter
                .map(|module| {
                    let mut bytes = Vec::new();
                    module.serialize(&mut bytes).unwrap();
                    (module.self_id().name().to_string(), bytes)
                })
                .collect(),
            max_move_package_size,
        )
    }

    /// Return the size of the package in bytes. Only count the bytes of the modules themselves--the
    /// fact that we store them in a map is an implementation detail
    pub fn size(&self) -> usize {
        self.module_map.values().map(|b| b.len()).sum()
    }

    pub fn id(&self) -> ObjectID {
        self.id
    }

    pub fn version(&self) -> SequenceNumber {
        self.version
    }

    /// Approximate size of the package in bytes. This is used for gas metering.
    pub fn object_size_for_gas_metering(&self) -> usize {
        // + 8 for version
        self.serialized_module_map()
            .iter()
            .map(|(name, module)| name.len() + module.len())
            .sum::<usize>()
            + 8
    }

    pub fn serialized_module_map(&self) -> &BTreeMap<String, Vec<u8>> {
        &self.module_map
    }

    pub fn deserialize_module(&self, module: &Identifier) -> SuiResult<CompiledModule> {
        // TODO use the session's cache
        let bytes = self
            .serialized_module_map()
            .get(module.as_str())
            .ok_or_else(|| SuiError::ModuleNotFound {
                module_name: module.to_string(),
            })?;
        Ok(CompiledModule::deserialize(bytes)
            .expect("Unwrap safe because Sui serializes/verifies modules before publishing them"))
    }

    pub fn disassemble(&self) -> SuiResult<BTreeMap<String, Value>> {
        disassemble_modules(self.module_map.values())
    }

    pub fn normalize(&self) -> SuiResult<BTreeMap<String, normalized::Module>> {
        normalize_modules(self.module_map.values())
    }
}

pub fn disassemble_modules<'a, I>(modules: I) -> SuiResult<BTreeMap<String, Value>>
where
    I: Iterator<Item = &'a Vec<u8>>,
{
    let mut disassembled = BTreeMap::new();
    for bytecode in modules {
        let module = CompiledModule::deserialize(bytecode).map_err(|error| {
            SuiError::ModuleDeserializationFailure {
                error: error.to_string(),
            }
        })?;
        let view = BinaryIndexedView::Module(&module);
        let d = Disassembler::from_view(view, Spanned::unsafe_no_loc(()).loc).map_err(|e| {
            SuiError::ObjectSerializationError {
                error: e.to_string(),
            }
        })?;
        let bytecode_str = d
            .disassemble()
            .map_err(|e| SuiError::ObjectSerializationError {
                error: e.to_string(),
            })?;
        disassembled.insert(module.name().to_string(), Value::String(bytecode_str));
    }
    Ok(disassembled)
}

pub fn normalize_modules<'a, I>(modules: I) -> SuiResult<BTreeMap<String, normalized::Module>>
where
    I: Iterator<Item = &'a Vec<u8>>,
{
    let mut normalized_modules = BTreeMap::new();
    for bytecode in modules {
        let module = CompiledModule::deserialize(bytecode).map_err(|error| {
            SuiError::ModuleDeserializationFailure {
                error: error.to_string(),
            }
        })?;
        let normalized_module = normalized::Module::new(&module);
        normalized_modules.insert(normalized_module.name.to_string(), normalized_module);
    }
    Ok(normalized_modules)
}
