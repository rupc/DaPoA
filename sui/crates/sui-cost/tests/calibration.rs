// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::path::PathBuf;

use insta::assert_snapshot;
use move_cli::base::reroot_path;
use move_disassembler::disassembler::Disassembler;
use move_package::BuildConfig;

const TEST_MODULE_DATA_DIR: &str = "../sui-framework/tests";

// Execute every entry function in Move framework and examples and ensure costs don't change
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta (`cargo install cargo-insta`)
// 1. Run `cargo insta test --review` under `./sui-cost`.
// 2. Review, accept or reject changes.

#[tokio::test]
async fn test_natives_disassemble_snapshot() -> Result<(), anyhow::Error> {
    let natives_calib = disassemble_test_module("natives_calibration_tests".to_string())?;

    // Assert that nothing has changed in the disassemby of the code
    assert_snapshot!(natives_calib);
    Ok(())
}

#[tokio::test]
async fn test_bytecode_disassemble_snapshot() -> Result<(), anyhow::Error> {
    let bytecode_calib = disassemble_test_module("bytecode_calibration_tests".to_string())?;

    // Assert that nothing has changed in the disassemby of the code
    assert_snapshot!(bytecode_calib);
    Ok(())
}

fn disassemble_test_module(name: String) -> anyhow::Result<String> {
    let path = PathBuf::from(TEST_MODULE_DATA_DIR);
    let config = BuildConfig {
        test_mode: true,
        ..Default::default()
    };
    let package_name: Option<String> = None;
    let module_name = name;

    let rerooted_path = reroot_path(Some(path))?;

    let package = config.compile_package(&rerooted_path, &mut Vec::new())?;
    let needle_package = package_name
        .as_deref()
        .unwrap_or(package.compiled_package_info.package_name.as_str());
    match package
        .get_module_by_name(needle_package, &module_name)
        .ok()
    {
        None => anyhow::bail!(
            "Unable to find module or script with name '{}' in package '{}'",
            module_name,
            needle_package,
        ),
        Some(unit) => Disassembler::from_unit(&unit.unit).disassemble(),
    }
}
