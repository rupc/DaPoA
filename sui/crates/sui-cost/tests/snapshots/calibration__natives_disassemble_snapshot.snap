---
source: crates/sui-cost/tests/calibration.rs
expression: natives_calib
---
// Move bytecode v6
module 2.natives_calibration_tests {
use 0000000000000000000000000000000000000000000000000000000000000001::unit_test;
use 0000000000000000000000000000000000000000000000000000000000000002::event;
use 0000000000000000000000000000000000000000000000000000000000000002::object;
use 0000000000000000000000000000000000000000000000000000000000000002::test_scenario;
use 0000000000000000000000000000000000000000000000000000000000000002::tx_context;


struct StructSimple has copy, drop, store {
	dummy_field: bool
}
struct ObjectWithID has store, key {
	id: UID
}

public calibrate_emit(obj#0#0: StructSimple) {
B0:
	0: MoveLoc[0](obj#0#0: StructSimple)
	1: Call event::emit<StructSimple>(StructSimple)
	2: Ret
}
public calibrate_emit_nop(obj#0#0: StructSimple) {
B0:
	0: MoveLoc[0](obj#0#0: StructSimple)
	1: Pop
	2: Ret
}
entry public test_calibrate_event_emit() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(15)
B2:
	6: Branch(7)
B3:
	7: LdFalse
	8: Pack[0](StructSimple)
	9: Call calibrate_emit(StructSimple)
	10: MoveLoc[0](trials#1#0: u64)
	11: LdU64(1)
	12: Sub
	13: StLoc[0](trials#1#0: u64)
	14: Branch(2)
B4:
	15: Ret
}
entry public test_calibrate_event_emit__baseline() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(15)
B2:
	6: Branch(7)
B3:
	7: LdFalse
	8: Pack[0](StructSimple)
	9: Call calibrate_emit_nop(StructSimple)
	10: MoveLoc[0](trials#1#0: u64)
	11: LdU64(1)
	12: Sub
	13: StLoc[0](trials#1#0: u64)
	14: Branch(2)
B4:
	15: Ret
}
entry public test_calibrate_id_address_from_bytes() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(14)
B2:
	6: Branch(7)
B3:
	7: LdConst[1](Vector(U8): [32, 50, 134, 111, 1, 9, 250, 27, 169, 17, 57, 45, 205, 45, 66, 96, 241, 216, 36, 49, 51, 22, 245, 70, 122, 191, 100, 24, 123, 62, 239, 165, 85])
	8: Call object::calibrate_address_from_bytes(vector<u8>)
	9: MoveLoc[0](trials#1#0: u64)
	10: LdU64(1)
	11: Sub
	12: StLoc[0](trials#1#0: u64)
	13: Branch(2)
B4:
	14: Ret
}
entry public test_calibrate_id_address_from_bytes__baseline() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(14)
B2:
	6: Branch(7)
B3:
	7: LdConst[1](Vector(U8): [32, 50, 134, 111, 1, 9, 250, 27, 169, 17, 57, 45, 205, 45, 66, 96, 241, 216, 36, 49, 51, 22, 245, 70, 122, 191, 100, 24, 123, 62, 239, 165, 85])
	8: Call object::calibrate_address_from_bytes_nop(vector<u8>)
	9: MoveLoc[0](trials#1#0: u64)
	10: LdU64(1)
	11: Sub
	12: StLoc[0](trials#1#0: u64)
	13: Branch(2)
B4:
	14: Ret
}
entry public test_calibrate_id_borrow_uid() {
L0:	obj#1#0: ObjectWithID
L1:	scenario#1#0: &mut Scenario
L2:	scenario_val#1#0: Scenario
L3:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[3](trials#1#0: u64)
	2: LdConst[2](Address: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
	3: Call test_scenario::begin(address): Scenario
	4: StLoc[2](scenario_val#1#0: Scenario)
	5: MutBorrowLoc[2](scenario_val#1#0: Scenario)
	6: StLoc[1](scenario#1#0: &mut Scenario)
B1:
	7: CopyLoc[3](trials#1#0: u64)
	8: LdU64(0)
	9: Gt
	10: BrFalse(27)
B2:
	11: Branch(12)
B3:
	12: CopyLoc[1](scenario#1#0: &mut Scenario)
	13: Call test_scenario::ctx(&mut Scenario): &mut TxContext
	14: Call object::new(&mut TxContext): UID
	15: Pack[1](ObjectWithID)
	16: StLoc[0](obj#1#0: ObjectWithID)
	17: ImmBorrowLoc[0](obj#1#0: ObjectWithID)
	18: Call object::calibrate_borrow_uid<ObjectWithID>(&ObjectWithID)
	19: MoveLoc[0](obj#1#0: ObjectWithID)
	20: Unpack[1](ObjectWithID)
	21: Call object::delete(UID)
	22: MoveLoc[3](trials#1#0: u64)
	23: LdU64(1)
	24: Sub
	25: StLoc[3](trials#1#0: u64)
	26: Branch(7)
B4:
	27: MoveLoc[1](scenario#1#0: &mut Scenario)
	28: Pop
	29: MoveLoc[2](scenario_val#1#0: Scenario)
	30: Call test_scenario::end(Scenario): TransactionEffects
	31: Pop
	32: Ret
}
entry public test_calibrate_id_borrow_uid__baseline() {
L0:	obj#1#0: ObjectWithID
L1:	scenario#1#0: &mut Scenario
L2:	scenario_val#1#0: Scenario
L3:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[3](trials#1#0: u64)
	2: LdConst[2](Address: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
	3: Call test_scenario::begin(address): Scenario
	4: StLoc[2](scenario_val#1#0: Scenario)
	5: MutBorrowLoc[2](scenario_val#1#0: Scenario)
	6: StLoc[1](scenario#1#0: &mut Scenario)
B1:
	7: CopyLoc[3](trials#1#0: u64)
	8: LdU64(0)
	9: Gt
	10: BrFalse(29)
B2:
	11: Branch(12)
B3:
	12: CopyLoc[1](scenario#1#0: &mut Scenario)
	13: Call test_scenario::ctx(&mut Scenario): &mut TxContext
	14: Call object::new(&mut TxContext): UID
	15: Pack[1](ObjectWithID)
	16: StLoc[0](obj#1#0: ObjectWithID)
	17: ImmBorrowLoc[0](obj#1#0: ObjectWithID)
	18: Call object::calibrate_borrow_uid<ObjectWithID>(&ObjectWithID)
	19: ImmBorrowLoc[0](obj#1#0: ObjectWithID)
	20: Pop
	21: MoveLoc[0](obj#1#0: ObjectWithID)
	22: Unpack[1](ObjectWithID)
	23: Call object::delete(UID)
	24: MoveLoc[3](trials#1#0: u64)
	25: LdU64(1)
	26: Sub
	27: StLoc[3](trials#1#0: u64)
	28: Branch(7)
B4:
	29: MoveLoc[1](scenario#1#0: &mut Scenario)
	30: Pop
	31: MoveLoc[2](scenario_val#1#0: Scenario)
	32: Call test_scenario::end(Scenario): TransactionEffects
	33: Pop
	34: Ret
}
entry public test_calibrate_tx_context_derive_id() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(15)
B2:
	6: Branch(7)
B3:
	7: LdConst[3](Vector(U8): [32, 58, 152, 93, 167, 79, 226, 37, 178, 4, 92, 23, 45, 107, 211, 144, 189, 133, 95, 8, 110, 62, 157, 82, 91, 70, 191, 226, 69, 17, 67, 21, 50])
	8: LdU64(0)
	9: Call tx_context::calibrate_derive_id(vector<u8>, u64)
	10: MoveLoc[0](trials#1#0: u64)
	11: LdU64(1)
	12: Sub
	13: StLoc[0](trials#1#0: u64)
	14: Branch(2)
B4:
	15: Ret
}
entry public test_calibrate_tx_context_derive_id__baseline() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(15)
B2:
	6: Branch(7)
B3:
	7: LdConst[3](Vector(U8): [32, 58, 152, 93, 167, 79, 226, 37, 178, 4, 92, 23, 45, 107, 211, 144, 189, 133, 95, 8, 110, 62, 157, 82, 91, 70, 191, 226, 69, 17, 67, 21, 50])
	8: LdU64(0)
	9: Call tx_context::calibrate_derive_id_nop(vector<u8>, u64)
	10: MoveLoc[0](trials#1#0: u64)
	11: LdU64(1)
	12: Sub
	13: StLoc[0](trials#1#0: u64)
	14: Branch(2)
B4:
	15: Ret
}
entry public test_calibrate_pop() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(14)
B2:
	6: Branch(7)
B3:
	7: LdFalse
	8: Pop
	9: MoveLoc[0](trials#1#0: u64)
	10: LdU64(1)
	11: Sub
	12: StLoc[0](trials#1#0: u64)
	13: Branch(2)
B4:
	14: Ret
}
entry public test_calibrate_pop__baseline() {
L0:	trials#1#0: u64
B0:
	0: LdConst[0](U64: [1, 0, 0, 0, 0, 0, 0, 0])
	1: StLoc[0](trials#1#0: u64)
B1:
	2: CopyLoc[0](trials#1#0: u64)
	3: LdU64(0)
	4: Gt
	5: BrFalse(12)
B2:
	6: Branch(7)
B3:
	7: MoveLoc[0](trials#1#0: u64)
	8: LdU64(1)
	9: Sub
	10: StLoc[0](trials#1#0: u64)
	11: Branch(2)
B4:
	12: Ret
}
unit_test_poison() {
B0:
	0: LdU64(0)
	1: Call unit_test::create_signers_for_testing(u64): vector<signer>
	2: Pop
	3: Ret
}
}
